tools/src/main/java/org/apache/kafka/trogdor/workload/RandomComponentPayloadGenerator.java

```diff
@@ -0,0 +1,113 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.trogdor.workload;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+
+/**
+ * A PayloadGenerator which generates pseudo-random payloads based on other PayloadGenerators.
+ *
+ * Given a seed and non-null list of RandomGeneratorConfigs, RandomComponentPayloadGenerator
+ * will use any given generator in its list of components a percentage of the time based on the 
+ * percent field in the RandomGeneratorConfig. These percent fields must be integers greater than 0 
+ * and together add up to 100. The payloads generated can be reproduced from run to run.
+ * 
+ * An example of how to include this generator in a Trogdor taskSpec is shown below.
+ * #{@code
+ *    "keyGenerator": {
+ *        "type": "randomComponent",
+ *        "seed": 456,
+ *        "components": [
+ *          {
+ *            "percent": 50,
+ *            "component": {
+ *              "type": "null"
+ *            }
+ *          },
+ *          {
+ *            "percent": 50,
+ *            "component": {
+ *              "type": "uniformRandom",
+ *              "size": 4,
+ *              "seed": 123,
+ *              "padding": 0
+ *            }
+ *          }
+ *        ]
+ *    }
+ * }
+ */
+public class RandomComponentPayloadGenerator implements PayloadGenerator {
+    private final long seed;
+    private final List<RandomGeneratorConfig> components;
+    private final Random random = new Random();
+
+    @JsonCreator
+    public RandomComponentPayloadGenerator(@JsonProperty("seed") long seed,
+                                           @JsonProperty("components") List<RandomGeneratorConfig> components) {
+        this.seed = seed;
+        if (components == null || components.isEmpty()) {
+            throw new IllegalArgumentException("Components must be a specified, non-empty list of RandomGeneratorConfigs.");
+        } else {
+            int sum = 0;
+            for (RandomGeneratorConfig generator : components) {
+                if (generator.percent() < 1) {
+                    throw new IllegalArgumentException("Percent value must be greater than zero.");
+                }
+                sum += generator.percent();
+            }
+            if (sum != 100) {
+                throw new IllegalArgumentException("Components must be a list of RandomGeneratorConfigs such that the percent fields sum to 100");
+            }
+        }
+        this.components = new ArrayList<>(components);
+
+    }
+
+    @JsonProperty
+    public long seed() {
+        return seed;
+    }
+
+    @JsonProperty
+    public List<RandomGeneratorConfig> components() {
+        return components;
+    }
+
+    @Override
+    public synchronized byte[] generate(long position) {
```

Comment:
we only need the lock for setting the seed and calling `random.nextInt`, right?  for the rest of the function we can avoid holding the lock since we're just iterating over an immutable list that can't change and calling stuff that is threadsafe anyway

Reply:
So I should remove the synchronized on the function and just make a block for the first two lines?

Reply:
yeah